### ISSUE TITLE
Fast-LoRA autograd: single-projection LoRA Function (o_proj, and the fallback for any other adapted Linear)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#839]
Part of #792 (fast-LoRA autograd tracker). The flag, validators, tolerance rule, NF4 rule, checkpointing and streaming criteria live there; this issue is the single-projection math and its own tests.

## Why

`o_proj` (transformers 5.17.0, `models/llama/modeling_llama.py:280`) is the one attention projection whose input — the attention output — no other projection shares, so there is nothing to concatenate. What is left against peft (`peft/tuners/lora/layer.py:1058`; for 4-bit, `bnb.py:533` and `:549`) is small: fold the scaling into the GEMM (`addmm(..., alpha=s)`), accumulate the LoRA term into the base output instead of allocating the LoRA output, its scaled copy and the sum (plus the 4-bit `clone()`), and accumulate `dX` into one buffer. The GEMM count does not change: three in forward, five in backward.

So the expected gain is small and may be zero. This issue is still worth doing, and is the natural first PR of the tracker, because it is the smallest math on which to establish the plumbing the other two reuse: instance-level patching, `save_for_backward` for every tensor (the #331 rule), NF4 dequantisation in backward, checkpoint and streaming behaviour, and delegation to peft under `disable_adapters`. It is also the fallback that lets a mixed `target_modules` list run through one mechanism.

## Math

peft/torch convention: `W: [out, in]`, `A: [r, in]`, `B: [out, r]`, `X: [N, in]`, `s` = the module's scaling. Dropout 0 (tracker rule).

Forward:

```

H = X @ A^T                         # [N, r]

Y = X @ W^T (+ b)                   # base; bias frozen

Y = addmm(Y, H, B^T, alpha=s)       # LoRA term accumulated in place of three allocations

```

Saved: `X`, `A`, `B`, and `W` (or packed NF4 + quant-state tensors) through `save_for_backward`; `H` (`[N, r]`) saved or recomputed — pick one and justify it with the measurement.

Backward, given `dY: [N, out]`:

```

dB = s * dY^T @ H                   # [out, r]

dH = s * dY @ B                     # [N, r]

dA = dH^T @ X                       # [r, in]

dX = dY @ W                         # NF4: dequantise W here, from saved tensors

dX = addmm(dX, dH, A)               # one buffer

```

## Stretch: dropout > 0 on this path only

With no shared X, dropout costs nothing structurally here, and there is a memory argument for it: peft's `lora_A` saves the full-precision dropped copy `X_d`, while this Function can save the boolean mask (1 byte per element) and rebuild `X_d = X * mask / (1 - p)` in backward:

```

dA = dH^T @ X_d

dX = dY @ W + (dH @ A) * mask / (1 - p)

```

Two conditions before the tracker's `dropout == 0` rule is relaxed for this path: under checkpoint recompute the same mask must come back (checkpoint restores RNG state by default; a test must prove the masks match), and parity with peft is only exact if the mask is drawn in the same RNG order as `nn.Dropout` — otherwise test with an injected fixed mask plus a statistical test of the drop rate. Until both hold, the validator keeps refusing.

## Scope

In: `o_proj`; any adapted `nn.Linear` or `bnb.nn.Linear4bit` not covered by the MLP or QKV paths, including fused `qkv_proj` / `gate_up_proj` Linears; frozen base bias; NF4. Out: GPT-2 `Conv1D` (`fan_in_fan_out` transposes the weight), embedding LoRA, `lm_head`, 8-bit.

## Tests (in addition to the tracker's shared set)

- [ ] gradcheck, float64, CPU: `N=5, in=8, out=6, r=2`, w.r.t. X, A, B; with and without base bias.

- [ ] Forward and backward parity vs an unpatched peft `lora.Linear` and `lora.bnb.Linear4bit`, reported separately. Where the Function issues the same ops as peft in the same order, say whether fp32/bf16 come out bit-exact; where `addmm` fusion breaks it, report the ratio per the tracker rule.

- [ ] A #331-shaped streaming test: more layers than `stream_buffers`, NF4, gradient on every layer's `o_proj` adapter compared, not only the last layers.

- [ ] `disable_adapters` and `merged` delegate to the original forward.

- [ ] Micro-benchmark at the Llama-3.1-8B `o_proj` shape (4096 -> 4096, `N=2048`, `r` in `{16, 64}`), bf16 and NF4: median forward+backward time and `torch.cuda.max_memory_allocated`, fast vs plain, with card, driver, SM clock and versions.

- [ ] Stretch: mask-equality under checkpoint and dropout parity with an injected mask.

## Hardware needed

CPU for gradcheck and fp32 parity. Any CUDA GPU for bf16/NF4 parity, the streaming test and the micro-benchmark.

## Honest limits

- On its own this path may show no measurable speed or memory difference, and that would be a fine outcome to post. Its value is the plumbing and, later, dropout.

- bf16 parity is tolerance-based unless the op sequence matches peft's exactly.

---

**Related, filed the same day:**

- #792 — Fast-LoRA autograd, correctness half: hand-written backward for the LoRA projections, behind training.fast_...

**Comment by sloemo01** [comment]
Taking this. I can cover the CI-able half: the single-projection Function with `torch.autograd.gradcheck` in float64 on CPU, fp32 forward and backward parity against the unpatched peft module, the `disable_adapters` case, and the fp32 loss curve, all wired into pytest.

No CUDA box on my side, so the bf16 / NF4 / streamed-suite items from the tracker will land as skip-in-CI tests until someone with a GPU can run them on the named card. I will state that in the PR rather than claim the numbers.

**Comment by MakazhanAlpamys** [comment]
The single-projection Function and its tests landed in #1087 (merged as 7b78e528), by @sloemo01 — thank you.

Against the checklist above, the merged `tests/test_issue839_fast_lora_single_projection.py` covers the float64 gradcheck with and without a base bias, fp32 parity against an unpatched peft `lora.Linear`, NF4 parity against unpatched peft (CUDA-gated), `disable_adapters` / `merged` delegation, and a streamed-vs-resident comparison on a 4-layer tiny Llama that checks the loss and all 16 LoRA gradients. The kernel is not wired to a flag yet; that is #792's step.

**Staying open on three items:**

- **The #331-shaped streaming test in NF4.** The merged streamed comparison is unquantized, and the item asks for NF4 with more layers than `stream_buffers`.
- **The micro-benchmark** at the Llama-3.1-8B `o_proj` shape (4096 -> 4096, `N=2048`, `r` in `{16, 64}`, bf16 and NF4), with median forward+backward time, `torch.cuda.max_memory_allocated`, and the card, driver, SM clock and versions. The merged benchmark test deliberately reports and asserts nothing, so a recorded run is still needed.
- The dropout stretch item.

**Comment by MakazhanAlpamys** [comment]
@sloemo01, a question about the three items still open here, because two of them need a CUDA card and on 09-19 you said you don't have one:

- The dropout stretch item (mask equality under checkpoint recompute, and parity with peft) can be done on CPU. Do you want it?
- The NF4 streaming test (more layers than `stream_buffers`) and the `o_proj` micro-benchmark need a card, so they are open to anyone who has one.

If you'd rather stop at #1087, say so and I'll unassign you, so the issue doesn't look reserved. Thank you again for #1087.

**Comment by sloemo01** [comment]
@MakazhanAlpamys Yes, I'll take it. Mask equality under checkpoint recompute and the dropout parity with an injected mask are both CPU work, so I'll cover those and open a small PR against this issue. The NF4 streaming test and the o_proj micro-benchmark stay open for whoever has a card. Thanks for the clear split.

**Comment by MakazhanAlpamys** [comment]
@sloemo01, great, the dropout stretch item is yours: mask equality under checkpoint recompute and dropout parity with an injected mask, both on CPU. I've marked the issue `claimed`. The NF4 streaming test and the `o_proj` micro-benchmark stay open for anyone with a CUDA card, so please use `Refs #839` rather than `Closes` in your PR.
